client/state/products-list/selectors/is-marketplace-product.js

```diff
@@ -0,0 +1,13 @@
+import 'calypso/state/products-list/init';
+
+const productSlugToSubscriptionSlug = ( productSlug ) => productSlug.replace( '-', '_' );
+
+export const isMarketplaceProduct = ( state, productSlug ) =>
```

Comment:
This may be more than you want to do, but the fact that you had to use a bound selector in the above file makes me think that this function is just doing too much.

What if you made a function (maybe in a different place than this file; could be where the original one was?) that looked like:

```js
export const isMarketplaceProduct = ( products, productSlug ) => ...
```

Then if you need to call it with the redux state, it's just one additional step:

```js
const products = getProductsList(); // getProductsList already exists to get `state.productsList.items`
const isProductMarketplace = isMarketPlaceProduct( products, productSlug );
```

The advantage is that then you could call `isMarketPlaceProduct()` without needing the full state on hand. So instead of the bound selector above you could do:

```js
// ... in connect()...
			productList: getProductsList( state )
// ... in the component ...
wasMarketplaceProduct = purchases.some( ( product ) => isMarketPlaceProduct( this.props.productList, product.productSlug ) );
```

Reply:
Checking a little bit more around this while working on https://github.com/Automattic/wp-calypso/issues/55575. @gcsecsey we can remove all marketplace references from `client/my-sites/checkout/checkout-thank-you/index.jsx` as there is another way to:

- install the marketplace product (it's done on subscription activation on the backend)
- redirect to any page if we want through a `redirect_to` parameter.

meaning that we (currently) don't care about a bound selector.

Reply:
BTW, @sirbrillig why do you feel this selector is doing too much? Adding a `productList: getProductsList( state )` dependency wherever we need `isMarketplaceProduct` seems too much to me.

Reply:
Thanks @sirbrillig, this is a brilliant point! I really like the idea of separating the function from the state, makes the usage much simpler.

Reply:
One question about moving the `isMarketPlaceProduct` to its original place. The products list schema is already defined as a JSON schema, but its elements' schema does not match the product interface in `/packages/calypso-products/src/types.ts`:
https://github.com/Automattic/wp-calypso/blob/39307f83e8128787b35c74e1d5d75da4ef27316d/client/state/products-list/schema.js#L1-L28

https://github.com/Automattic/wp-calypso/blob/430b075bc47a59b10de28c8d883b3926a365563e/packages/calypso-products/src/types.ts#L68-L77

@sirbrillig Do you think the existing interface could be extended, or should I create a new interface for this schema?

Reply:
> Checking a little bit more around this while working on #55575. @gcsecsey we can remove all marketplace references from client/my-sites/checkout/checkout-thank-you/index.jsx as there is another way to:
> 
>  - install the marketplace product (it's done on subscription activation on the backend)
>  - redirect to any page if we want through a redirect_to parameter.
> 
> meaning that we (currently) don't care about a bound selector.

Sorry @cpapazoglou I've just seen your comments regarding this. Do you mean that the loading of this component can also be removed?
https://github.com/Automattic/wp-calypso/blob/4994d76671ecdd0a16a7efa279f0a58d439ad996/client/my-sites/checkout/checkout-thank-you/index.jsx#L459-L462

That would make the bound selector the easier implementation IMO, as it wasn't used anywhere else.

Reply:
yep

Reply:
> why do you feel this selector is doing too much? Adding a productList: getProductsList( state ) dependency wherever we need isMarketplaceProduct seems too much to me.

There's already at least one place where you're ending up with a leaky abstraction, and the selector is doing two things: getting data from state, then processing that data. If those operations are separated, it becomes a tool that's useful outside of the context of state. Plus it allows for composition.

You're already performing an implicit query to the products list, so it wouldn't be more actual processing. Certainly it adds a prop that you may not otherwise need in the component, but there are other reasons for getting the product list that some (future or existing) components may have; this way they could share that state. There are already a fair number of places that do this.

But the best part is composition. If you make the function more decoupled from the data selection, then you could still build a combined selector if you like! The best of both worlds. So you could have:

```js
export const isMarketplaceProduct = ( products, productSlug ) => ...
```

```js
export function isMarketPlaceProductInState( state, productSlug ) {
  const products = getProductsList( state );
  return isMarketPlaceProduct( products, productSlug );
}
```

> Do you think the existing interface could be extended, or should I create a new interface for this schema?

I think it could probably be extended. If that's one that I added (likely) I probably added just enough of it to cover whatever use-case I was trying to make pass TypeScript at the time, and it could definitely be missing some things.

Reply:
> Do you think the existing interface could be extended, or should I create a new interface for this schema?

Is it the same type of data? I mean, is that `Product` the same type of `Product` found in the products list? I ask because it appears to have functions like `getProductId` which I imagine don't exist in state. If they do match, then feel free to add to that `Product` definition. Otherwise, I suggest making a new type. 

IMO it never really hurts to make a new type if you get a more accurate definition of what you need on an object. Since TS is mostly structural, it allows for things like this to work:

```ts
interface RegularProduct {
  product_name: string;
  product_id: number;
}

interface UnrelatedProduct {
  product_name: string;
  getID: () => number;
}

interface SomethingWithAName {
  product_name: string;
}

function getProductName( product: { product_name: string } ): string {
  return product.product_name;
}
```

Reply:
In this case specifically, it looks like the type you actually need is `{ product_type: string; product_slug: string }[]`.

Reply:
Thanks @sirbrillig for all your great points and help! 🎉 

> There's already at least one place where you're ending up with a leaky abstraction, and the selector is doing two things: getting data from state, then processing that data. If those operations are separated, it becomes a tool that's useful outside of the context of state.

I've updated the selector to work with the different suffixes and separators in product slugs that we currently have in state. With this string manipulation needed, I think the data processing is inevitably bound to the schema, but separating the data fetching can still make the usage of the selector easier from both class- and functional components.

I've moved the processing part back and its tests to its original place, so that it can be used outside of the context of state. I've also added the `getProductsList` dependency directly in the `isMarketplaceProduct` selector, so that it can be used without adding the dependency each time.

I'm still figuring out ways to test regressions because the original function was used in more places, I'll update the testing instructions with these soon.